Lib/Rabbit4000/tcpip/DMAETH100.LIB

```diff
@@ -1020,72 +1040,122 @@ netport_tx_isr::
+   jr	nz,.netport_done_notxinc	; don't really care...and it shouldn't count as a transmitted packet
+; nothing else should come in here
    jr		.netport_done
 
 .tx_packet_error:
-ioi ld   a,(NBTSR)
+; Transmitted packet in error.  Check NBTSR and NBTESR for details
+; Some errors might have aborted the packet where we do not get the dma tx interrupt
+;  in that case, we need to do some special stuff
+	ld	b,0				; the aborted packet flag
 #ifdef DMAETH100_NET_STATS
-   ; Transmitted packet in error.  Check NBTSR for details
    ld    hl,(ix+[_decb]+pd_tx_error)
    inc   hl
    ld    (ix+[_decb]+pd_tx_error),hl
-
+#endif
+ioi ld   a,(NBTSR)
    bit   4,a
    jr    z,.txe1
+   ld	b,1				; packet aborted for underrun
+#ifdef DMAETH100_NET_STATS
    ld    hl,(ix+[_decb]+pd_txe_urun)
    inc   hl
    ld    (ix+[_decb]+pd_txe_urun),hl
-   jr    .txe3
+#endif
+   jr    .txeF
 .txe1:
    bit   5,a
    jr    z,.txe2
-	ld		b,1		; Reset aborted Tx DMA due to 16 collisions
+	ld	b,1		; packet aborted due to excessive collisions
+#ifdef DMAETH100_NET_STATS
    ld    hl,(ix+[_decb]+pd_txe_colls)
    inc   hl
    ld    (ix+[_decb]+pd_txe_colls),hl
-   jr    .txe3
+#endif
+   jr    .txeF
 .txe2:
-	ld		b,0		; For deferred Tx, don't reset Tx DMA
-   bit   6,a
+; the rest of the possible errors do not abort the transmission
+; AND we really only care about them if we have stats running
+	ld	b,0		; the rest of the possible errors do not abort the transmission
+#ifdef DMAETH100_NET_STATS
+	bit   6,a		; we count deferred packets
    jr    z,.txe3
    ld    hl,(ix+[_decb]+pd_txe_defer)
    inc   hl
    ld    (ix+[_decb]+pd_txe_defer),hl
-#endif
+   jr	.txeF
 .txe3:
+; look at NBTESR for some more transmit error sources
 ioi ld   a,(NBTESR)        ; 11
-#ifdef DMAETH100_NET_STATS
-   bit   4,a
-   jr    z,.netport_done
+   bit   4,a		; check for crc error
+   jr    z,.txeF
    ld    hl,(ix+[_decb]+pd_txe_crc)
    inc   hl
    ld    (ix+[_decb]+pd_txe_crc),hl
 #endif
+.txeF:
+; Ok, done categorizing errors
+	djnz	.netport_done
+; Frame was aborted...need to halt dma if needed and clean up a bit
+	ld   a, (ix+[_decb]+sending)
+	cp	0
+	jr	z,.txePurgeOnly
+; tx dma interrupt didn't get called...so clean it up here
+#if _RAB6K
+	ld   hl,(ix+[_decb]+txbit)	; Halt current DMA
+	ioi  ld   (DMHLR),hl
+#else
+	ld    a,(ix+[_decb]+txbit)	; Halt current DMA
+	ioi   ld   (DMHR),a
+#endif
+; Return buffer to packet buffer pool
+	ld    jkhl,(ix+[_decb]+txbuf)
+	ld    py,jkhl		; Buffer address to be freed
+	push	ix
+	lcall	_pb_free
+	pop		ix
+; Not sending any more
+	ld   (ix+[_decb]+sending),0
+.txePurgeOnly:
+; May as well clear the fifo too
+	ld 	a,0x10     	; Purge FIFO
+	ioi	ld (NBCR),a
+; not sure we should count this as "transmitted" for stats, but what the heck
 
 .netport_done:
 #ifdef DMAETH100_NET_STATS
    ld    hl,(ix+[_decb]+pd_transmitted)
    inc   hl
    ld    (ix+[_decb]+pd_transmitted),hl
 #endif
+.netport_done_notxinc:
+#if !defined(_RAB6K)

```

Comment:
Be sure to remove this if going back to old code where interrupt priority changes on both 5000 and 6000.